### ISSUE TITLE
Prepend support for bytes

### ISSUE THREAD

**Issue by jq-rs** [carllerche/bytes#161]
Hi! Thanks for this great crate! While trying to avoid unnecessary copying it would be a nice feature to be able to insert data to the beginning of a byte array, i.e. prepend a large set of data with a small header. I could not figure out how to do it at the moment. AFAI can understand, both chaining and extending introduce an extra alloc. Is it possible to do somehow or would we need some kind of prepend() and with_capacity() with prepend max size option?

**Comment by jq-rs** [comment]
I think I figured it out. When header is received, the BytesMut can be increased to proper size with reserve() and split_off() to header and message parts where message is passed to the read socket. When message is received, the header can be extended with the received message without allocation.

**Comment by jq-rs** [comment]
In the end, it seems that there is no way extending Bytes without new allocation after split_off. The idea would be to combine small and large buffer together without copy (small.extend(large)), after large has been received from io::read_exact().

**Comment by carllerche** [comment]
Do you need the header & payload to be in contiguous memory? If not, you can use `Buf::chain` to represent them in discontinuous memory.

**Comment by jq-rs** [comment]
Chain is good as such, but when I collect(), it will do an alloc and copy, I think.

I can easily arrange the buffers to be continuous, so I think if extend() would check ptr + len fields for continues memory, then new alloc would be avoided. Just modify then small len to include large, and drop large. No need for a new prepend API then.

**Comment by carllerche** [comment]
I'm not quite sure what you are trying to achieve.

**Comment by jq-rs** [comment]
Sorry for the confusion, I'll try to explain more: I am trying to avoid copy of large-frame when I receive it and extend it to small-frame. The large-frame can be several MBs as small one is 16B, thus I really would not like to copy it when small.extend(large).

I was thinking about a function like (pseudocode, probably does not compile nor is exactly correct):
```rust
pub fn unsplit(&mut self, BytesMut: other) -> BytesMut {
    if self.inner.ptr + self.inner.len == other.inner.ptr { //continuous blocks, just combine directly
       self.inner.len += other.inner.len;
       drop(other);
   }
   else {
        self.extend(other);
   }
   self
}
```

If there would be a way to pass just part of a BytesMut array to read_exact(), then this would not be needed. Like `read_exact(reader, &bytes[starting from this location])`. AFAIK, it is not possible.

**Comment by jq-rs** [comment]
4c01e81925c116ed98f94612858b6573be18d30a has a fix proposal for this.

**Comment by jq-rs** [comment]
The fix #162 is now reviewed so this issue can be closed as soon as the fix is merged. Thanks!

**Comment by jq-rs** [comment]
Fix 2ca61d8 merged
